### PR TITLE
(PDB-1996) Remove HSQLDB support

### PR DESCRIPTION
**Pull request by rbrw**

### CODE REVIEW

File: test/puppetlabs/puppetdb/http/facts_test.clj [hunk 1]
```diff
@@ -581,74 +580,6 @@
           (is (= body msg))
           (is (= status http/status-bad-request)))))))
 
-(deftestseq ^{:postgres false} two-database-fact-query-config
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/http/facts_test.clj]: @wkalt Do we want to preserve this test? Maybe we could move it to the sync repo where two-databases is already assumed.

Comment (wkalt) [inline test/puppetlabs/puppetdb/http/facts_test.clj]: I think we need to port the two-database handling to FOSS so we can preserve this. We need coverage of the two database scenario in core as well.

Comment (rbrw) [inline test/puppetlabs/puppetdb/http/facts_test.clj]: Agreed.  I'll see what's involved.

**Comment by rbrw** [comment]
This will be WIP until https://github.com/puppetlabs/puppetdb/pull/1663 and https://github.com/puppetlabs/puppetdb/pull/1665 make it to master.

File: ext/travisci/test.sh [hunk 2]
```diff
@@ -8,27 +8,16 @@ ulimit -u 4096
...
-        PUPPETDB_DBTYPE=hsqldb lein2 test
-        ;;
-      *)
-        echo "Invalid database: $PDB_TEST_DB" 1>&2
-        exit 1
-        ;;
-    esac
+    psql -c 'create database puppetdb_test;' -U postgres
+    psql -c 'create database puppetdb2_test;' -U postgres
+    PUPPETDB_DBTYPE=postgres \
```
Comment (ajroetker) [inline ext/travisci/test.sh]: Don't think we need this setting anymore :)

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 3]
```diff
@@ -983,11 +953,11 @@
...
   ;;   Changing the types of hashes and uuids in postgres to bytea and uuid respectively
   []
-  (let [hash-type (if (sutils/postgres?) "bytea" "varchar(40)")
-        uuid-type (if (sutils/postgres?) "uuid" "varchar(255)")
-        json-type (if (sutils/postgres?) "json" "text")
-        munge-hash (if (sutils/postgres?) (fn [column] (format "('\\x' || %s)::bytea" column)) identity)
-        munge-uuid (if (sutils/postgres?) (fn [column] (format "%s::uuid" column)) identity)]
+  (let [hash-type "bytea"
+        uuid-type "uuid"
+        json-type "json"
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: These don't really need to be bound names anymore

File: src/puppetlabs/puppetdb/scf/storage_utils.clj [hunk 4]
```diff
@@ -185,60 +173,26 @@ must be supplied as the value to be matched."
...
   [column]
   [:and
    [(keyword "~") column "?"]
    [:is-not column nil]])
 
-(defmethod sql-regexp-match "HSQL Database Engine"
-  [column]
-  [:is-not (h/regexp-substring column "?") nil])
-
 (defmethod sql-regexp-array-match "PostgreSQL"
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/storage_utils.clj]: These don't need to be defmethods anymore right?

File: test/puppetlabs/puppetdb/http/reports_test.clj [hunk 5]
```diff
@@ -74,13 +75,13 @@
       (is (= (query-result method endpoint ["extract" "logs"
                                             ["=" "certname" (:certname basic)]])
              #{{:logs (merge {:href (utils/as-path "/v4/reports" report-hash "logs")}
-                             (when (sutils/postgres?) {:data (get-in basic [:logs :data])}))}})))
+                             {:data (get-in basic [:logs :data])})}})))
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/http/reports_test.clj]: I think this can just go into the other map that's in the `merge` args.

File: test/puppetlabs/puppetdb/http/reports_test.clj [hunk 6]
```diff
@@ -74,13 +75,13 @@
...
 
     (testing "metrics projected"
       (is (= (query-result method endpoint ["extract" "metrics"
                                             ["=" "certname" (:certname basic)]])
              #{{:metrics (merge {:href (utils/as-path "/v4/reports" report-hash "metrics")}
-                                (when (sutils/postgres?) {:data (get-in basic [:metrics :data])}))}})))
+                                {:data (get-in basic [:metrics :data])})}})))
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/http/reports_test.clj]: Same note here.

File: test/puppetlabs/puppetdb/scf/migrate_test.clj [hunk 7]
```diff
@@ -410,9 +398,7 @@
                          INNER JOIN report_statuses rs ON r.status_id=rs.id
                          ORDER BY c.certname"
                         (sutils/sql-hash-as-str "r.hash")
-                        (sutils/sql-uuid-as-str "r.transaction_uuid")
-                        cast-to-jsonb
-                        cast-to-jsonb))]
+                        (sutils/sql-uuid-as-str "r.transaction_uuid")))]
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/scf/migrate_test.clj]: We can probably just get rid of this function and just use `::text` directly in the SQL like you have for `jsonb`

File: config.sample.ini [hunk 8]
```diff
@@ -17,20 +17,9 @@ vardir = /var/lib/puppetdb
...
-# https://groups.google.com/d/msg/puppet-users/8K5sPqNgErM/8PI5pjI5iRgJ
-
-# For PostgreSQL: org.postgresql.Driver
-# For the embedded DB: org.hsqldb.jdbcDriver
+# Subname pattern: //host:port/databaseName
 #classname = org.postgresql.Driver
-
-# For PostgreSQL: postgresql
-# For the embedded DB: hsqldb
 #subprotocol = postgresql
```
Comment (ajroetker) [inline config.sample.ini]: Should we just set these two settings above?

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 3 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: Oh, right - I converted this code first, then just started replacing these values in the SQL, but didn't get back to these.

File: documentation/configure.markdown [hunk 9]
```diff
@@ -333,25 +314,22 @@ If unset, the default value is 14 days.
...
+change it.
 
 ### `subprotocol`
 
-Set this to:
-
-* `hsqldb` when using the embedded database
-* `postgresql` when using PostgreSQL
+This should be `postgresql`, which is the default.  You should not
+need to change it..
```
Comment (senior) [inline documentation/configure.markdown]: extra period

File: documentation/install_from_packages.markdown [hunk 10]
```diff
@@ -61,18 +60,8 @@ Use Puppet to install PuppetDB:
...
-- For the moment, PuppetDB also supports an embedded HSQLDB database,
-  but this support has been deprecated and will be removed in a future
-  release. Please see [this email][hsqldb_deprecation_mail] to the
-  puppet-users list for further information.  If you do choose HSQLDB,
-  you should [increase PuppetDB's maximum heap size][configure_heap]
-  to at least 1 GB.
-
-You can change PuppetDB's database at any time, but note that changing the database does not migrate PuppetDB's data, so the new database will be empty. However, as this data is automatically generated many times a day, PuppetDB should recover in a relatively short period of time.
+- [Set up a PostgreSQL server and configure PuppetDB to use it][configure_postgres].
+  You may also need to [adjust the maximum heap size][configure_heap].
```
Comment (senior) [inline documentation/install_from_packages.markdown]: Does this configure_heap part need to be there? Seems like it's not related to configuring the database anymore

File: documentation/scaling_recommendations.markdown [hunk 11]
```diff
@@ -23,25 +22,14 @@ As with scaling any service, there are several possible performance and reliabil
...
-
 ### PostgreSQL Speed and Availability
 
-Using the PostgreSQL backend, PuppetDB will be limited by the performance of your Postgres server. You can increase performance by making sure your DB server has an extremely fast disk, plenty of RAM, a fast processor, and a fast network connection to your PuppetDB server. You may also need to look into database clustering and load balancing.
+PuppetDB will be limited by the performance of your Postgres
+server. You can increase performance by making sure your DB server has
+an extremely fast disk, plenty of RAM, a fast processor, and a fast
+network connection to your PuppetDB server. You may also need to look
+into database clustering and load balancing.  Note that the default
+PostgreSQL configuration on your system may be very conservative.
```
Comment (senior) [inline documentation/scaling_recommendations.markdown]: @wkalt seems like we should get pg_tune into our scaling recommendations, is it something we can just put a paragraph on, or would it be a significant effort (i.e. separate page)?

File: src/puppetlabs/puppetdb/cli/services.clj [hunk 12]
```diff
@@ -192,8 +192,8 @@
 (defn initialize-schema
   "Ensure the database is migrated to the latest version, and warn if
   it's deprecated, log and exit if it's unsupported. We do this in a
-  single connection because HSQLDB seems to get confused if the
-  database doesn't exist but we open and close a connection without
+  single connection because HSQLDB seemed to get confused if the
```
Comment (senior) [inline src/puppetlabs/puppetdb/cli/services.clj]: is this still relevant?

File: src/puppetlabs/puppetdb/config.clj [hunk 13]
```diff
@@ -54,8 +54,8 @@
      :conn-max-age (pls/defaulted-maybe s/Int 60)
      :conn-keep-alive (pls/defaulted-maybe s/Int 45)
      :conn-lifetime (s/maybe s/Int)
-     :classname (s/maybe String)
-     :subprotocol (s/maybe String)
+     :classname (pls/defaulted-maybe String "org.postgresql.Driver")
+     :subprotocol (pls/defaulted-maybe String "postgresql")
```
Comment (senior) [inline src/puppetlabs/puppetdb/config.clj]: :+1:

File: src/puppetlabs/puppetdb/config.clj [hunk 14]
```diff
@@ -157,36 +157,27 @@
...
-          msg (str "database configuration is now required."
-                   " Previously HSQLDB was the default, but support has been deprecated."
-                   " If you were relying on that default,"
-                   " please add this to the [database] section of your config file:"
-                   (setting-needed :classname)
-                   (setting-needed :subprotocol)
-                   (setting-needed :subname))]
+  [{db-config :database :or {db-config {}} :as config}]
+  (let [bad-opt
+        (when (or (not= "org.postgresql.Driver" (:classname db-config))
```
Comment (senior) [inline src/puppetlabs/puppetdb/config.clj]: Why do we need this?

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 15]
```diff
@@ -49,7 +48,7 @@
   [column-keyword]
   (->> column-keyword
        name
-       su/sql-uuid-as-str
+       (#(str % "::text"))
```
Comment (senior) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: It would be great if this had a name, would be more clear what appending ::text does

Comment (ajroetker) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: Also I think this could be turned into
```
(-> column-keyword name (str "::text"))
```

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 16]
```diff
@@ -318,10 +316,10 @@
   []
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: Also, none of these `query-rec`'s need to be functions now that we're always using postgresql.

File: src/puppetlabs/puppetdb/scf/storage_utils.clj [hunk 17]
```diff
@@ -252,59 +186,29 @@ must be supplied as the value to be matched."
...
-    db-uuid))
+  (.toString db-uuid))
 
 (pls/defn-validated parse-db-json
   "Produce a function for parsing an object stored as json."
   [db-json :- (s/maybe (s/either s/Str PGobject))]
-  (if-let [json (if (postgres?)
-                  (when db-json (.getValue db-json))
-                  db-json)]
+  (when-let [json (when db-json (.getValue db-json))]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/storage_utils.clj]: I think this is all just `(some-> db-json (.getValue) (json/parse-string true)`

File: test/puppetlabs/puppetdb/scf/storage_test.clj [hunk 18]
```diff
@@ -804,10 +804,12 @@
...
     (testing "changing a resource title"
-      (let [{orig-id :id
-             orig-tx-id :transaction_uuid
-             orig-timestamp :timestamp}
-            (first (query-to-vec (format "SELECT id, timestamp, transaction_uuid%s from catalogs where certname=?" (if (sutils/postgres?) "::text" "")) certname))
+      (let [[{orig-id :id
+              orig-tx-id :transaction_uuid
+              orig-timestamp :timestamp}]
+            (query-to-vec "select id, timestamp, transaction_uuid::text
+                             from catalogs where certname=?"
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/scf/storage_test.clj]: Might wanna split this up using `str` now that it's newline separated

File: test/puppetlabs/puppetdb/scf/storage_test.clj [hunk 19]
```diff
@@ -849,9 +851,10 @@
                                    FROM catalogs c INNER JOIN catalog_resources cr ON c.id = cr.catalog_id
                                    WHERE c.certname=?" certname))))
 
-        (let [results (query-to-vec (format "SELECT id, timestamp, %s from catalogs where certname=?"
-                                            (sutils/sql-uuid-as-str "transaction_uuid"))
-                                    certname)
+        (let [results (query-to-vec
+                       "select id, timestamp, transaction_uuid::text
+                          from catalogs where certname=?"
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/scf/storage_test.clj]: same here

File: test/puppetlabs/puppetdb/config_test.clj [hunk 20]
```diff
@@ -57,30 +57,31 @@
...
-                                   cfg-rx
-                                   (validate-db-settings {x "foo"})))
-            ;; Make sure the message doesn't contain a setting for x.
-            (is (try+
-                 (validate-db-settings {:database {x "foo"}})
-                 true
-                 (catch [:type ::conf/cli-error] {:keys [message]}
-                   (not (re-find (re-pattern (str " " (name x) " = "))
-                                 message)))))))))
+    (let [cfg {:database {:classname "org.postgresql.Driver"
```
Comment (ajroetker) [inline test/puppetlabs/puppetdb/config_test.clj]: Could we also add some tests for the HSQLDB case just, would be helpful just to know what we expect to happen?

File: src/puppetlabs/puppetdb/scf/storage_utils.clj [hunk 21]
```diff
@@ -252,59 +186,29 @@ must be supplied as the value to be matched."
...
-           ;; so here we return 1 if the maxid is nil, and otherwise return
-           ;; maxid +1 to indicate that the next number should be higher
-           ;; then the current one.
-           restartid (if (nil? maxid) 1 (inc maxid))]
-       (jdbc/do-commands (str "ALTER TABLE " table " ALTER COLUMN " column
-                              " RESTART WITH " restartid))))))
+    (jdbc/do-commands (str "LOCK TABLE " table " IN ACCESS EXCLUSIVE MODE"))
+    (jdbc/query-with-resultset
+     [(str "SELECT setval(pg_get_serial_sequence(?, ?),
+                          (SELECT max(" column ") FROM " table "))")
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/storage_utils.clj]: I think the string can be broken up a little more so that there isn't all the extra whitespace in the query.

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 22]
```diff
@@ -1611,8 +1556,7 @@
 (defn indexes!
   "Create missing indexes for applicable database platforms."
   [config]
-  (if (and (sutils/postgres?)
-           (sutils/db-version-newer-than? [9 2]))
+  (if (sutils/db-version-newer-than? [9 2])
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: I'm not sure we need this check at all anymore, since we only support 9.4+.

File: src/puppetlabs/puppetdb/query_eng/engine.clj [hunk 23]
```diff
@@ -752,42 +752,29 @@
...
-                       :factsets :certname
-                       :reports :hash
-                       nil)]
-    (if (or expand? (not expandable?))
-      [field name]
-      [href-key name])))
+  "Returns all fields from a projection.  Returns nil for fields which
+  are query-only? since these can't be projected either."
+  [[name {:keys [field]}] entity]
+  [field name])
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: This seems like it's really just `(juxt first (comp :field second))`

File: src/puppetlabs/puppetdb/query_eng/engine.clj (hunk 16 above)
Comment (senior) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: We can fix that up later imo

**Comment by senior** [comment]
I'm a :+1: with the minor comments above

File: documentation/scaling_recommendations.markdown (hunk 11 above)
Comment (wkalt) [inline documentation/scaling_recommendations.markdown]: @senior yeah I agree completely. I think we could mention it with a link here and then think about creating a page as a separate task. IIRC there's an outstanding docs ticket around postgres tuning since it affects more than just us, but I couldn't find it last time I looked.

Comment (rbrw) [inline documentation/scaling_recommendations.markdown]: @wkalt Sounds good (I'd wondered about that too, i.e. if we wanted to endorse pgtune).  I'll add the link and a brief comment for now.

File: src/puppetlabs/puppetdb/cli/services.clj (hunk 12 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/cli/services.clj]: @senior I'd left it so that we'd know why the code was structured the way it is (if/when we come back along to rework things - which, given the System/exit, I'm hoping we do).  I could also just move that text from the doc string to a comment - or were you suggesting I should go ahead and rework the code now?

File: src/puppetlabs/puppetdb/query_eng/engine.clj (hunk 23 above)
Comment (rbrw) [inline src/puppetlabs/puppetdb/query_eng/engine.clj]: @ajroetker hmm, not sure yet, but I might prefer the destructuring since the names make it a bit clearer to me, though I might also be happy with `[[name info] entity] [(:field info) name]` if that's preferable.

File: test/puppetlabs/puppetdb/config_test.clj (hunk 20 above)
Comment (rbrw) [inline test/puppetlabs/puppetdb/config_test.clj]: @ajroetker what did you have in mind?  Since HSQLDB support's gone, I think it should (currently) be treated just like any other invalid setting, and so may be covered by the "Bad :classname or :subprotocol" test.

Comment (ajroetker) [inline test/puppetlabs/puppetdb/config_test.clj]: Well for our other retired settings we special case them such that we give a nice warning saying, hey this setting is retired go do this thing instead.

**Comment by senior** [comment]
:+1: - I made one small comment on moving the two flushes to a flush-all function or something, but I see no reason to hold up the PR on that, that can be addressed later, maybe int he 2053 ticket you just created or something. Let me know when you get the new commits collapsed and I'll merge it in.

File: .travis.yml [hunk 24]
```diff
@@ -8,8 +8,7 @@ jdk:
   - oraclejdk7
 
 env:
-  - PDB_TEST_LANG=clojure PDB_TEST_DB=hsqldb
-  - PDB_TEST_LANG=clojure PDB_TEST_DB=postgres
```
Comment (ajroetker) [inline .travis.yml]: Sorry to be late to the game here, but fyi you'll have to remove this from the jenkins config as well `ext/jenkins/lein-test.sh`

**Comment by ajroetker** [comment]
I'm +1 here, good to merge once tests pass

**Comment by rbrw** [comment]
@puppetlabs-jenkins test this please
